File: server/src/main/java/com/tp/holdem/server/HoldemServerListener.java
package com.tp.holdem.server;

import com.esotericsoftware.kryonet.Connection;
import com.esotericsoftware.kryonet.Listener;
import com.tp.holdem.logic.PlayerExceptions;
import com.tp.holdem.logic.model.Player;
import com.tp.holdem.logic.model.PokerTable;
import com.tp.holdem.model.common.Phase;
import com.tp.holdem.model.message.Message;
import com.tp.holdem.model.message.MessageType;
import com.tp.holdem.model.message.PlayerActionMessage;
import com.tp.holdem.model.message.PlayerConnectMessage;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;

@Slf4j
@RequiredArgsConstructor
class HoldemServerListener extends Listener {
	private final MessageSender sender;
	private final GameParams params;
	private final GameHandler gameHandler;

	private ConnectedPlayers connectedPlayers = ConnectedPlayers.empty();
	private int expectActionFrom;

	@Override
	public synchronized void received(final Connection connection, final Object object) {
		if (object instanceof Message) {
			log.debug(String.format("Received message from %d: %s", connection.getID(), object));

			if (connection.getID() != expectActionFrom)
				throw new IllegalStateException("Unexpected player sent action");

			final Integer playerNumber = connectedPlayers.getConnected(connection.getID())
					.getOrElseThrow(PlayerExceptions.PLAYER_NOT_FOUND);

			final Message message = (Message) object;

			if (message.getMessageType() == MessageType.PLAYER_ACTION) {
				log.debug(String.format("Received action of type %s from player %d", message.getMessageType(), playerNumber));

				final PlayerActionMessage content = message.instance(PlayerActionMessage.class);

				PokerTable response = gameHandler.handlePlayerMove(playerNumber, content);

				if (response.isShowdown()) {
					log.debug("Starting showdown");

					while (response.getPhase() != Phase.RIVER) {
						response = gameHandler.startPhase();

						sender.sendStateUpdate(connectedPlayers, response);
						expectActionFrom = findExpectedResponder(response);

						try {
							Thread.sleep(1500);
						} catch (InterruptedException e) {
							log.debug("Waiting interrupted");
						}
					}

					response = gameHandler.roundOver();
				}

				sender.sendStateUpdate(connectedPlayers, response);
				expectActionFrom = findExpectedResponder(response);

				if (response.getPhase() == Phase.OVER) {
					log.debug("Current round is over, sleeping for 5s and staring new round");

					try {
						Thread.sleep(5000);
					} catch (InterruptedException e) {
						log.debug("Sleep interrupted");
					}

					startRound();
				}
			}
		}
	}

	@Override
	public synchronized void connected(final Connection con) {
		log.debug(String.format("Connection attempt: %d", con.getID()));

		final boolean connected = tryConnectingPlayer(con);

		if (!connected) {
			log.debug("Could not connect player");
			return;
		}

		if (enoughPlayers()) {
			log.debug("Enough players connected, starting game");
			startGame();
		}
	}

	@Override
	public synchronized void disconnected(final Connection con) {
		log.debug(String.format("Dis-connection attempt: %d", con.getID()));

		if (!connectedPlayers.isConnected(con.getID()))
			throw new IllegalStateException("Player is not connected");

		connectedPlayers.getConnected(con.getID())
				.map(gameHandler::disconnectPlayer)
				.forEach(table -> sender.sendStateUpdate(connectedPlayers, table));
	}

	private boolean tryConnectingPlayer(Connection con) {
		if (enoughPlayers()) {
			log.debug("Already enough players, could not connect");

			final PlayerConnectMessage response = PlayerConnectMessage.failure();
			sender.sendSingle(con.getID(), Message.from(MessageType.PLAYER_CONNECTION, response));
			return false;
		}

		if (connectedPlayers.isConnected(con.getID()))
			throw new IllegalStateException("Player already connected");

		final Player player = gameHandler.connectPlayer();
		connectedPlayers = connectedPlayers.connect(con.getID(), player.getNumber());

		log.debug(String.format("Connected player: %d", player.getNumber()));

		final PlayerConnectMessage connectionResponse = PlayerConnectMessage.success(player.toCurrentPlayerDTO());
		sender.sendSingle(con.getID(), Message.from(MessageType.PLAYER_CONNECTION, connectionResponse));
		return true;
	}

	private void startGame() {
		log.debug("Staring new game");

		final PokerTable response = gameHandler.startGame();
		sender.sendStateUpdate(connectedPlayers, response);
		expectActionFrom = findExpectedResponder(response);
	}

	private void startRound() {
		log.debug("Starting new round");

		final PokerTable response = gameHandler.startRound();
		sender.sendStateUpdate(connectedPlayers, response);
		expectActionFrom = findExpectedResponder(response);
	}

	private int findExpectedResponder(PokerTable response) {
		if (response.getPhase() == Phase.OVER)
			return -1;

		return response
				.getBettingPlayer()
				.map(Player::getNumber)
				.flatMap(connectedPlayers::getConnectionId)
				.getOrElse(-1);
	}

	private boolean enoughPlayers() {
		return connectedPlayers.size() >= params.getPlayerCount();
	}
}

File: server/src/main/kotlin/com/tp/holdem/server/HoldemServerListener.kt
package com.tp.holdem.server

import com.esotericsoftware.kryonet.Connection
import com.esotericsoftware.kryonet.Listener
import com.tp.holdem.common.lazyLogger
import com.tp.holdem.common.message.Message
import com.tp.holdem.common.message.MessageType
import com.tp.holdem.common.message.PlayerActionMessage
import com.tp.holdem.common.message.PlayerConnectMessage
import com.tp.holdem.common.model.Phase
import com.tp.holdem.logic.model.PlayerNumber
import com.tp.holdem.logic.utils.toCurrentPlayerDTO
import com.tp.holdem.logic.model.PokerTable
import com.tp.holdem.logic.players.number
import com.tp.holdem.logic.table.currentPhase
import com.tp.holdem.logic.table.currentlyBetting
import com.tp.holdem.logic.table.inShowdownMode

internal class HoldemServerListener(
        private val sender: MessageSender,
        private val params: GameParams,
        private val gameHandler: GameHandler
) : Listener() {
    private var connectedPlayers = ConnectedPlayers.empty()
    private var expectActionFrom = PlayerNumber.empty()

    private val log by lazyLogger()

    @Synchronized
    override fun received(connection: Connection, message: Any) {
        if (message is Message) {
            log.debug("Received message from ${connection.id}: $message")

            if (connection.id != expectActionFrom.number)
                throw IllegalStateException("Unexpected player sent action")

            val playerNumber = connectedPlayers.getConnected(connection.id)
                    .getOrElseThrow { IllegalStateException("Player not found") }

            if (message.messageType == MessageType.PLAYER_ACTION) {
                log.debug("Received action of type ${message.messageType} from player $playerNumber")

                val content = message.instance(PlayerActionMessage::class.java)

                var response = gameHandler.handlePlayerMove(playerNumber, content)

                if (response.inShowdownMode()) {
                    response = handleShowdown(response)
                }

                sender.sendStateUpdate(connectedPlayers, response)
                        .also { expectActionFrom = findExpectedResponder(response) }

                if (response.currentPhase() == Phase.OVER) {
                    waitAndStartNewRound()
                }
            }
        }
    }

    private fun handleShowdown(table: PokerTable): PokerTable {
        var response = table
        log.debug("Starting showdown")

        while (response.currentPhase() != Phase.RIVER) {
            response = gameHandler.startPhase()
                    .also { sender.sendStateUpdate(connectedPlayers, response) }
                    .also { expectActionFrom = findExpectedResponder(response) }
                    .also { Thread.sleep(1500) }
        }

        return gameHandler.roundOver()
    }

    private fun waitAndStartNewRound() {
        log.debug("Current round is over, sleeping for 5s and staring new round")

        Thread.sleep(5000)
                .also { startRound() }
    }

    @Synchronized
    override fun connected(con: Connection) {
        log.debug("Connection attempt: ${con.id}")

        val connected = tryConnectingPlayer(con)

        if (!connected) {
            return log.debug("Could not connect player")
        }

        if (enoughPlayers()) {
            log.debug("Enough players connected, starting game")
            startGame()
        }
    }

    @Synchronized
    override fun disconnected(con: Connection) {
        log.debug("Disconnection attempt: ${con.id}")

        if (!connectedPlayers.isConnected(con.id))
            throw IllegalStateException("Player is not connected")

        connectedPlayers.getConnected(con.id)
                .map(gameHandler::disconnectPlayer)
                .forEach { table -> sender.sendStateUpdate(connectedPlayers, table) }
    }

    private fun tryConnectingPlayer(con: Connection): Boolean {
        if (enoughPlayers()) {
            return sender.sendSingle(con.id, Message.from(MessageType.PLAYER_CONNECTION, PlayerConnectMessage.failure()))
                    .also { log.debug("Already enough players, could not connect") }
                    .let { false }
        }

        if (connectedPlayers.isConnected(con.id))
            throw IllegalStateException("Player already connected")

        return gameHandler.connectPlayer()
                .also { connectedPlayers = connectedPlayers.connect(con.id, it.number()) }
                .also { log.debug("Connected player: ${it.number()}") }
                .also { sender.sendSingle(con.id, Message.from(MessageType.PLAYER_CONNECTION, PlayerConnectMessage.success(it.toCurrentPlayerDTO()))) }
                .let { true }
    }

    private fun startGame() {
        log.debug("Staring new game")

        gameHandler.startGame()
                .also { sender.sendStateUpdate(connectedPlayers, it) }
                .also { expectActionFrom = findExpectedResponder(it) }
    }

    private fun startRound() {
        log.debug("Starting new round")

        gameHandler.startRound()
                .also { sender.sendStateUpdate(connectedPlayers, it) }
                .also { expectActionFrom = findExpectedResponder(it) }
    }

    private fun findExpectedResponder(response: PokerTable): PlayerNumber =
            when {
                response.currentPhase() == Phase.OVER -> PlayerNumber.empty()
                else -> PlayerNumber.of(connectedPlayers.getConnectionId(response.currentlyBetting()).getOrElse(-1))
            }

    private fun enoughPlayers(): Boolean = connectedPlayers.size() >= params.playerCount
}
